drivers/charger/bq2407x.c

```diff
@@ -0,0 +1,203 @@
+{
+	const struct ti_bq2407x_config *config = dev->config;
+
+	if (gpio_pin_get_dt(&config->pg_pin)) {
+		val->online = CHARGER_ONLINE_FIXED;
+	} else {
+		val->online = CHARGER_ONLINE_OFFLINE;
+	}
+}
+
+static void ti_bq2407x_get_prop_status(const struct device *dev, union charger_propval *val)
+{
+	const struct ti_bq2407x_config *config = dev->config;
+	bool stat1;
+	bool stat2;
+
+	if (!gpio_pin_get_dt(&config->pg_pin)) {
+		/* External supply is not present */
+		val->status = CHARGER_STATUS_DISCHARGING;
+		return;
+	}
+
+	if (!gpio_pin_get_dt(&config->ce_pin)) {
+		/* Charging is not enabled */
+		val->status = CHARGER_STATUS_NOT_CHARGING;
+		return;
+	}
+
+	/*
+	 * The status table is as follows [stat1,stat2]
+	 *   [1,1] CHARGER_STATUS_CHARGING (precharge)
+	 *   [1,0] CHARGER_STATUS_CHARGING (fast charge)
+	 *   [0,1] CHARGER_STATUS_FULL
+	 *   [0,0] CHARGER_STATUS_NOT_CHARGING
+	 */
+	stat1 = gpio_pin_get_dt(&config->stat1_pin);
+	stat2 = gpio_pin_get_dt(&config->stat2_pin);
+	if ((stat1 && stat2) || stat1) {
+		val->status = CHARGER_STATUS_CHARGING;
+	} else if (stat2) {
+		val->status = CHARGER_STATUS_FULL;
+	} else {
+		val->status = CHARGER_STATUS_NOT_CHARGING;
+	}
+}
+
+static void ti_bq2407x_get_prop_charge_type(const struct device *dev, union charger_propval *val)
+{
+	const struct ti_bq2407x_config *config = dev->config;
+
+	if (gpio_pin_get_dt(&config->ce_pin)) {
+		val->charge_type = CHARGER_CHARGE_TYPE_ADAPTIVE;
+	} else {
+		val->charge_type = CHARGER_CHARGE_TYPE_NONE;
+	}
+}
+
+static int ti_bq2407x_get_property(const struct device *dev, const charger_prop_t prop,
+				   union charger_propval *val)
+{
+	struct ti_bq2407x_data *data = dev->data;
+	int ret = -ENOTSUP;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	switch (prop) {
+	case CHARGER_PROP_ONLINE:
+		ti_bq2407x_get_prop_online(dev, val);
+		ret = 0;
+		break;
+
+	case CHARGER_PROP_STATUS:
+		ti_bq2407x_get_prop_status(dev, val);
+		ret = 0;
+		break;
+
+	case CHARGER_PROP_CHARGE_TYPE:
+		ti_bq2407x_get_prop_charge_type(dev, val);
+		ret = 0;
+		break;
+
+	default:
+		break;
+	}
+
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static int ti_bq2407x_set_prop_charge_type(const struct device *dev,
+					   const union charger_propval *val)
+
+{
+	const struct ti_bq2407x_config *config = dev->config;
+
+	switch (val->charge_type) {
+	case CHARGER_CHARGE_TYPE_NONE:
+		/* Disable charging */
+		gpio_pin_set_dt(&config->ce_pin, 0);
+		break;
+
+	case CHARGER_CHARGE_TYPE_ADAPTIVE:
+		/* Enable charging */
+		gpio_pin_set_dt(&config->ce_pin, 1);
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int ti_bq2407x_set_property(const struct device *dev, const charger_prop_t prop,
+				    const union charger_propval *val)
+{
+	struct ti_bq2407x_data *data = dev->data;
+	int ret = -ENOTSUP;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	switch (prop) {
+	case CHARGER_PROP_CHARGE_TYPE:
```

Comment:
Is that documented somewhere? :)

Given the fact that you can seemingly set the status value to `CHARGER_STATUS_FULL` `CHARGER_STATUS_UNKNOWN` and that its named "status", I don't think this is intuitive?

Reply:
No, that's my bad. I can work on improving this documentation.

 This property is inherited from power_supply in Linux. From `linux/Documentation/ABI/testing/sysfs-class-power`:
```
What:		/sys/class/power_supply/<supply_name>/status
Date:		May 2007
Contact:	linux-pm@vger.kernel.org
Description:
		Represents the charging status of the battery. Normally this
		is read-only reporting although for some supplies this can be
		used to enable/disable charging to the battery.

		Access: Read, Write

		Valid values:
			      "Unknown", "Charging", "Discharging",
			      "Not charging", "Full"
```

I would like to maintain this common interpretation in Zephyr if that's alright with you.

Reply:
So making the driver accept setting "charging" and "not charging" to enable/disable charging then?

I usually agree with the way Linux does things, sysfs is not one of those things :) It is too flexible compared to dedicated APIs, leading to ambiguity and confusion as values within a single property themselves get properties (like r/w).

A dedicated charger_charge_enable(const struct device *dev, bool enabled); and having a charger_get_status(const struct device *dev, enum charger_status *status) with the "unknown" removed (the function can just return -ENOTSUP) would be unambiguous.

We are not limited by SYSFS like Linux is, everything is not a file on Zephyr, we can make actual functions that the linker can verify at build time :)

Reply:
Then this would require refactoring the existing max20335-charger driver.
See: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/charger/charger_max20335.c

Reply:
>So making the driver accept setting "charging" and "not charging" to enable/disable charging then?

correct

>A dedicated charger_charge_enable(const struct device *dev, bool enabled); and having a charger_get_status(const struct device *dev, enum charger_status *status) with the "unknown" removed (the function can just return -ENOTSUP) would be unambiguous.

I'm not opposed to this, it has minimal impact for my future plans for this API. @aaronemassey, any concerns with this wrt what your team will be working on in the near future? This seems like one of those things that we could do better Linux.

>We are not limited by SYSFS like Linux is, everything is not a file on Zephyr, we can make actual functions that the linker can verify at build time :)

This is valid.

Reply:
sorry for the delay @bjarki-trackunit, I went ahead and drafted up a PR for the dedicated charge enable handler here #66128. Let me know what you think there.